### DISCUSSION TITLE
Vehicle Routing Problem Time windows

### DISCUSSION THREAD

**Discussion by rahulj2409** [google/or-tools#2636] [Routing (and legacy CP) questions]
Python

I am wondering how to incorporate time windows for the Vehicle Routing Problem when the times are not "starting" from zero. IE my times are datetime timestamps ranging from random points in time from the month of february 2021. However in the example they give in the Google OR documentation, https://developers.google.com/optimization/routing/vrptw
There is a time window that starts at 0 and goes onwards. I was wondering if the time windows needs to necessarily start from zero for the problem to work.

Thanks

IE:
i have a bunch of time windows
[123, 456]
[567,789]
[45, 98],
etc..
where its all time starting from timestamp() as defined in python.

**Comment by Mizux** [comment]
You don't necessary need the TW to start at zero for any visit location.
For Start location you can also not start at zero **BUT** when creating the dimension using AddDimension() you must **not** force start cumul to zero aka boolean must be `false` !
https://github.com/google/or-tools/blob/b37d9c786b69128f3505f15beca09e89bf078a89/ortools/constraint_solver/routing.h#L451 

side note: the end of any TW must be less or equal to the vehicle max capacity set in the AddDimension() aka must be a valid domain otherwise solver will be in an "infeasible" state and instantly return no solution.

**Reply by rahulj2409** [reply]
Thanks!
Just to make sure, https://developers.google.com/optimization/routing/vrptw, in this example , the  fix_start_cumul_to_zero variable seems to be set to false when creating the travel time dimension. So I can essentially use the exact same code as this example, but with the time windows and time_matrix modified and it should work then?

What do you mean by vehicle max capacity in terms of the time window, are you talking about the "# maximum time per vehicle" constraint? If so, how do I ignore that, my goal is essentially utilize all of my vehicles in a equal manner, I don't want to have to focus on seeing the maximum time that each vehicle can spend.

Thanks!

**Comment by rahulj2409** [comment]
It errors at this line: 
time_dimension.CumulVar(index).SetRange(int(time_window[0]), int(time_window[1]))

it says "CP Solver fail", not really sure where I'm going wrong here

Thanks!
